File: plugins/binaryninja/analysis/src/PE/TypeBuilder.cpp
/* Copyright 2025 - 2026 R. Thomas
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "binaryninja/api_compat.hpp"
#include "binaryninja/analysis/PE/TypeBuilder.hpp"
#include "LIEF/PE/LoadConfigurations/LoadConfiguration.hpp"

using namespace BinaryNinja;

namespace analysis_plugin::pe {

Ref<Type> TypeBuilder::get_or_create(const std::string& name) {

  if (Ref<Type> type = get(name)) {
    return type;
  }

  const Ref<Architecture> default_arch = bv_.GetDefaultArchitecture();

  if (name == "IMAGE_ARM64EC_METADATA_V1") {
    StructureBuilder struct_builder;
    struct_builder
      .AddMember(u32(), "Version")
      .AddMember(RVA(), "CodeMap")
      .AddMember(u32(), "CodeMapCount")
      .AddMember(RVA(), "CodeRangesToEntryPoints")
      .AddMember(RVA(), "RedirectionMetadata")
      .AddMember(RVA(), "__os_arm64x_dispatch_call_no_redirect")
      .AddMember(RVA(), "__os_arm64x_dispatch_ret")
      .AddMember(RVA(), "__os_arm64x_dispatch_call")
      .AddMember(RVA(), "__os_arm64x_dispatch_icall")
      .AddMember(RVA(), "__os_arm64x_dispatch_icall_cfg")
      .AddMember(RVA(), "AlternateEntryPoint")
      .AddMember(RVA(), "AuxiliaryIAT")
      .AddMember(u32(), "CodeRangesToEntryPointsCount")
      .AddMember(u32(), "RedirectionMetadataCount")
      .AddMember(RVA(), "GetX64InformationFunctionPointer")
      .AddMember(RVA(), "SetX64InformationFunctionPointer")
      .AddMember(RVA(), "ExtraRFETable")
      .AddMember(u32(), "ExtraRFETableSize")
      .AddMember(RVA(), "__os_arm64x_dispatch_fptr")
      .AddMember(RVA(), "AuxiliaryIATCopy")
    ;
    Ref<Structure> S = struct_builder.Finalize();
    return create_struct(
        *S, "_" + format_type(name), format_type(name));
  }

  if (name == "IMAGE_ARM64EC_METADATA_V2") {
    Ref<Type> base = get_or_create("IMAGE_ARM64EC_METADATA_V1");
    auto base_struct = get_as<Structure>(base);
    assert(base_struct != nullptr);
    StructureBuilder builder;

    for (const StructureMember& member : base_struct->GetMembers()) {
      builder.AddMember(member.type, member.name, member.access, member.scope);
    }

    builder
      .AddMember(RVA(), "AuxDelayloadIAT")
      .AddMember(RVA(), "AuxDelayloadIATCopy")
      .AddMember(u32(), "ReservedBitField");

    Ref<Structure> S = builder.Finalize();
    return create_struct(
        *S, "_" + format_type(name), format_type(name));
  }

  if (name == "IMAGE_LOAD_CONFIG_CODE_INTEGRITY") {
    StructureBuilder builder;
    builder
      .AddMember(u16(), "Flags")
      .AddMember(u16(), "Catalog")
      .AddMember(u32(), "CatalogOffset")
      .AddMember(u32(), "Reserved");
    Ref<Structure> S = builder.Finalize();
    return create_struct(*S, format_type(name));
  }

  if (name == "IMAGE_GUARD") {
    using IMAGE_GUARD = LIEF::PE::LoadConfiguration::IMAGE_GUARD;
    EnumerationBuilder F;
    F.AddMemberWithValue(
        "IMAGE_GUARD_CF_INSTRUMENTED", (uint32_t)IMAGE_GUARD::CF_INSTRUMENTED);

    F.AddMemberWithValue(
        "IMAGE_GUARD_CFW_INSTRUMENTED", (uint32_t)IMAGE_GUARD::CFW_INSTRUMENTED);

    F.AddMemberWithValue(
        "IMAGE_GUARD_CF_FUNCTION_TABLE_PRESENT", (uint32_t)IMAGE_GUARD::CF_FUNCTION_TABLE_PRESENT);

    F.AddMemberWithValue(
        "IMAGE_GUARD_SECURITY_COOKIE_UNUSED", (uint32_t)IMAGE_GUARD::SECURITY_COOKIE_UNUSED);

    F.AddMemberWithValue(
        "IMAGE_GUARD_PROTECT_DELAYLOAD_IAT", (uint32_t)IMAGE_GUARD::PROTECT_DELAYLOAD_IAT);

    F.AddMemberWithValue(
        "IMAGE_GUARD_DELAYLOAD_IAT_IN_ITS_OWN_SECTION", (uint32_t)IMAGE_GUARD::DELAYLOAD_IAT_IN_ITS_OWN_SECTION);

    F.AddMemberWithValue(
        "IMAGE_GUARD_CF_EXPORT_SUPPRESSION_INFO_PRESENT", (uint32_t)IMAGE_GUARD::CF_EXPORT_SUPPRESSION_INFO_PRESENT);

    F.AddMemberWithValue(
        "IMAGE_GUARD_CF_ENABLE_EXPORT_SUPPRESSION", (uint32_t)IMAGE_GUARD::CF_ENABLE_EXPORT_SUPPRESSION);

    F.AddMemberWithValue(
        "IMAGE_GUARD_CF_LONGJUMP_TABLE_PRESENT", (uint32_t)IMAGE_GUARD::CF_LONGJUMP_TABLE_PRESENT);

    F.AddMemberWithValue(
        "IMAGE_GUARD_RF_INSTRUMENTED", (uint32_t)IMAGE_GUARD::RF_INSTRUMENTED);

    F.AddMemberWithValue(
        "IMAGE_GUARD_RF_ENABLE", (uint32_t)IMAGE_GUARD::RF_ENABLE);

    F.AddMemberWithValue(
        "IMAGE_GUARD_RF_STRICT", (uint32_t)IMAGE_GUARD::RF_STRICT);

    F.AddMemberWithValue(
        "IMAGE_GUARD_RETPOLINE_PRESENT", (uint32_t)IMAGE_GUARD::RETPOLINE_PRESENT);

    F.AddMemberWithValue(
        "IMAGE_GUARD_EH_CONTINUATION_TABLE_PRESENT", (uint32_t)IMAGE_GUARD::EH_CONTINUATION_TABLE_PRESENT);

    F.AddMemberWithValue(
        "IMAGE_GUARD_XFG_ENABLED", (uint32_t)IMAGE_GUARD::XFG_ENABLED);

    F.AddMemberWithValue(
        "IMAGE_GUARD_CASTGUARD_PRESENT", (uint32_t)IMAGE_GUARD::CASTGUARD_PRESENT);

    F.AddMemberWithValue(
        "IMAGE_GUARD_MEMCPY_PRESENT", (uint32_t)IMAGE_GUARD::MEMCPY_PRESENT);

    Ref<Enumeration> enum_ = F.Finalize();
    Ref<Type> enum_ty = Type::EnumerationType(
        default_arch, enum_, /*width=*/4, /*isSigned=*/false);

    return cache_.insert({format_type(name), std::move(enum_ty)}).first->second;
  }


  if (name == "IMAGE_ARM64EC_METADATA_CODE_RANGE") {
    StructureBuilder builder;
#if BN_BITFIELD_SUPPORT
    builder
      .SetPacked(true)
      .AddMemberAtBitOffset(u32(), "Type", /*bitOffset=*/0, /*bitWidth=*/2)
      .AddMemberAtBitOffset(u32(), "RVA", /*bitOffset=*/2, /*bitWidth=*/30)
    ;
#else
    builder
      .AddMember(u32(), "RVAType")
      .AddMember(u32(), "Length")
    ;
#endif


    Ref<Structure> S = builder.Finalize();
    return create_struct(
        *S, "_" + format_type(name), format_type(name));
  }

  if (name == "IMAGE_ARM64EC_CODE_RANGE_ENTRY_POINT") {
    StructureBuilder builder;
    builder
      .AddMember(RVA(), "StartRva")
      .AddMember(RVA(), "EndRva")
      .AddMember(RVA(), "EntryPoint")
    ;

    Ref<Structure> S = builder.Finalize();
    return create_struct(
        *S, "_" + format_type(name), format_type(name));
  }

  if (name == "IMAGE_ARM64EC_METADATA_REDIRECTION") {
    StructureBuilder builder;
    builder
      .AddMember(RVA(), "From")
      .AddMember(RVA(), "To")
    ;

    Ref<Structure> S = builder.Finalize();
    return create_struct(
        *S, "_" + format_type(name), format_type(name));
  }

  if (name == "IMAGE_ARM64_RUNTIME_FUNCTION_ENTRY") {
    StructureBuilder builder;
    builder
      .AddMember(RVA(), "BeginAddress")
      .AddMember(RVA(), "UnwindData");

    Ref<Structure> S = builder.Finalize();
    return create_struct(
        *S, "_" + format_type(name), format_type(name));
  }

  if (name == "IMAGE_ARM64_RUNTIME_FUNCTION_ENTRY_XDATA") {
    StructureBuilder builder;

#if BN_BITFIELD_SUPPORT
    builder
      .SetPacked(true)
      .AddMemberAtBitOffset(u32(), "FunctionLength", /*bitOffset=*/0, /*bitWidth=*/18)
      .AddMemberAtBitOffset(u32(), "Version", /*bitOffset=*/18, /*bitWidth=*/2)
      .AddMemberAtBitOffset(u32(), "ExceptionDataPresent", /*bitOffset=*/20, /*bitWidth=*/1)
      .AddMemberAtBitOffset(u32(), "EpilogInHeader", /*bitOffset=*/21, /*bitWidth=*/1)
      .AddMemberAtBitOffset(u32(), "EpilogCount", /*bitOffset=*/22, /*bitWidth=*/5)
      .AddMemberAtBitOffset(u32(), "CodeWords", /*bitOffset=*/27, /*bitWidth=*/5)
    ;
#else
    builder
      .AddMember(u32(), "HeaderData");
#endif

    Ref<Structure> S = builder.Finalize();
    return create_struct(*S, format_type(name));
  }

  if (name == "IMAGE_ARM64_RUNTIME_FUNCTION_ENTRY_XDATA_EXTENDED") {
    StructureBuilder builder;

#if BN_BITFIELD_SUPPORT
    builder
      .SetPacked(true)
      .AddMemberAtBitOffset(u32(), "ExtendedEpilogCount", /*bitOffset=*/0, /*bitWidth=*/16)
      .AddMemberAtBitOffset(u32(), "ExtendedCodeWords", /*bitOffset=*/16, /*bitWidth=*/8)
    ;
#else
    builder
      .AddMember(u32(), "ExtendedHeaderData");
#endif

    Ref<Structure> S = builder.Finalize();
    return create_struct(*S,format_type(name));
  }

  if (name == "IMAGE_ARM64_RUNTIME_FUNCTION_EXTENDED_ENTRY") {
    StructureBuilder builder;
    builder
      .AddMember(get_or_create("IMAGE_ARM64_RUNTIME_FUNCTION_ENTRY_XDATA"), "base")
      .AddMember(get_or_create("IMAGE_ARM64_RUNTIME_FUNCTION_ENTRY_XDATA_EXTENDED"), "extension")
    ;

    Ref<Structure> S = builder.Finalize();
    return create_struct(*S, format_type(name));
  }


  return nullptr;
}

}

File: plugins/ghidra/src/main/java/lief/ghidra/plugins/analyzers/pe/TypeBuilder.java
/***
 * Copyright 2022 - 2026 R. Thomas
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package lief.ghidra.plugins.analyzers.pe;

import java.util.Optional;

import ghidra.program.model.data.CategoryPath;
import ghidra.program.model.data.DataType;
import ghidra.program.model.data.DataTypeManager;
import ghidra.program.model.data.EnumDataType;
import ghidra.program.model.data.InvalidDataTypeException;
import ghidra.program.model.data.StructureDataType;
import ghidra.program.model.data.TypedefDataType;
import ghidra.program.model.data.UnionDataType;
import ghidra.util.Msg;

import lief.ghidra.plugins.analyzers.Context;

public class TypeBuilder<T> extends lief.ghidra.plugins.analyzers.TypeBuilder<T> {
    public final static DataType ULONG = U32;

    public TypeBuilder(Context<T> context) {
        super(context);
    }

    public void addBitField(StructureDataType S, DataType baseDataType,
        int bitSize, String componentName, String comment)
    {
        try {
            S.addBitField(baseDataType, bitSize, componentName, comment);
        } catch (InvalidDataTypeException e) {
            Msg.error(TypeBuilder.class, e);
        }
    }

    @Override
    public Optional<DataType> getType(String name) {
        DataTypeManager typeManager = context
            .getProgram()
            .getDataTypeManager();

        switch (name) {
            case "IMAGE_LOAD_CONFIG_CODE_INTEGRITY": {
                StructureDataType S =
                    new StructureDataType("_IMAGE_LOAD_CONFIG_CODE_INTEGRITY", 0);
                S.add(U16, "Flags",
                    "Flags to indicate if CI information is available, etc."
                );
                S.add(U16, "Catalog",
                    "0xFFFF means not available"
                );
                S.add(U32, "CatalogOffset",
                    "0xFFFF means not available"
                );
                S.add(U32, "Reserved",
                    "Additional bitmask to be defined later"
                );
                S.setCategoryPath(getCategoryPath());

                TypedefDataType typedef =
                    new TypedefDataType("IMAGE_LOAD_CONFIG_CODE_INTEGRITY", S);
                return Optional.of(
                    typeManager.addDataType(typedef, DEFAULT_TYPE_CONFLICT_HANDLER)
                );
            }

            case "IMAGE_GUARD_FLAGS": {
                EnumDataType E =
                    new EnumDataType("IMAGE_GUARD_FLAGS", U32.getLength());

                E.add("IMAGE_GUARD_CF_INSTRUMENTED", 0x00000100,
                    "Module performs control flow integrity checks using system-supplied support"
                );

                E.add("IMAGE_GUARD_CFW_INSTRUMENTED", 0x00000200,
                    "Module performs control flow and write integrity checks"
                );

                E.add("IMAGE_GUARD_CF_FUNCTION_TABLE_PRESENT", 0x00000400,
                    "Module contains valid control flow target metadata"
                );

                E.add("IMAGE_GUARD_SECURITY_COOKIE_UNUSED", 0x00000800,
                    "Module does not make use of the /GS security cookie"
                );

                E.add("IMAGE_GUARD_PROTECT_DELAYLOAD_IAT", 0x00001000,
                    "Module supports read only delay load IAT"
                );

                E.add("IMAGE_GUARD_DELAYLOAD_IAT_IN_ITS_OWN_SECTION", 0x00002000,
                    "Delayload import table in its own .didat section (with nothing else in it) that can be freely reprotected"
                );

                E.add("IMAGE_GUARD_CF_EXPORT_SUPPRESSION_INFO_PRESENT", 0x00004000,
                    "Module contains suppressed export information. This also infers that the address taken" +
                    "taken IAT table is also present in the load config"
                );

                E.add("IMAGE_GUARD_CF_ENABLE_EXPORT_SUPPRESSION", 0x00008000,
                    "Module enables suppression of exports"
                );

                E.add("IMAGE_GUARD_CF_LONGJUMP_TABLE_PRESENT", 0x00010000,
                    "Module contains longjmp target information"
                );

                E.add("IMAGE_GUARD_RF_INSTRUMENTED", 0x00020000,
                    "Module contains return flow instrumentation and metadata"
                );

                E.add("IMAGE_GUARD_RF_ENABLE", 0x00040000,
                    "Module requests that the OS enable return flow protection"
                );

                E.add("IMAGE_GUARD_RETPOLINE_PRESENT", 0x00100000,
                    "Module was built with retpoline support"
                );

                E.add("IMAGE_GUARD_EH_CONTINUATION_TABLE_PRESENT", 0x00400000,
                    "Module contains EH continuation target information"
                );

                E.add("IMAGE_GUARD_XFG_ENABLED", 0x00800000,
                    "Module was built with xfg (deprecated)"
                );

                E.add("IMAGE_GUARD_CASTGUARD_PRESENT", 0x01000000,
                    "Module has CastGuard instrumentation present"
                );

                E.add("IMAGE_GUARD_MEMCPY_PRESENT", 0x02000000,
                    "Module has Guarded Memcpy instrumentation present"
                );

                E.setCategoryPath(getCategoryPath());

                return Optional.of(
                    typeManager.addDataType(E, DEFAULT_TYPE_CONFLICT_HANDLER)
                );
            }

            case "IMAGE_ARM64EC_METADATA": {
                StructureDataType S =
                    new StructureDataType("_IMAGE_ARM64EC_METADATA", 0);
                S.add(ULONG, "Version", null);
                S.add(RVA,   "CodeMap", null);
                S.add(ULONG, "CodeMapCount", null);
                S.add(RVA,   "CodeRangesToEntryPoints", null);
                S.add(RVA,   "RedirectionMetadata", null);
                S.add(RVA,   "__os_arm64x_dispatch_call_no_redirect", null);
                S.add(RVA,   "__os_arm64x_dispatch_ret", null);
                S.add(RVA,   "__os_arm64x_dispatch_call", null);
                S.add(RVA,   "__os_arm64x_dispatch_icall", null);
                S.add(RVA,   "__os_arm64x_dispatch_icall_cfg", null);
                S.add(RVA,   "AlternateEntryPoint", null);
                S.add(RVA,   "AuxiliaryIAT", null);
                S.add(ULONG, "CodeRangesToEntryPointsCount", null);
                S.add(ULONG, "RedirectionMetadataCount", null);
                S.add(RVA,   "GetX64InformationFunctionPointer", null);
                S.add(RVA,   "SetX64InformationFunctionPointer", null);
                S.add(RVA,   "ExtraRFETable", null);
                S.add(ULONG, "ExtraRFETableSize", null);
                S.add(RVA,   "__os_arm64x_dispatch_fptr", null);
                S.add(RVA,   "AuxiliaryIATCopy", null);
                S.setCategoryPath(getCategoryPath());

                TypedefDataType typedef =
                    new TypedefDataType("IMAGE_ARM64EC_METADATA", S);
                return Optional.of(
                    typeManager.addDataType(typedef, DEFAULT_TYPE_CONFLICT_HANDLER)
                );
            }

            case "IMAGE_ARM64EC_METADATA_V2": {
                DataType base = getType("IMAGE_ARM64EC_METADATA").get();
                StructureDataType S =
                    new StructureDataType("_IMAGE_ARM64EC_METADATA_V2", 0);
                S.add(base, "V1", null);

                S.add(RVA,   "AuxDelayloadIAT", null);
                S.add(RVA,   "AuxDelayloadIATCopy", null);
                S.add(ULONG, "ReservedBitField", null);

                S.setCategoryPath(getCategoryPath());

                TypedefDataType typedef =
                    new TypedefDataType("IMAGE_ARM64EC_METADATA_V2", S);
                return Optional.of(
                    typeManager.addDataType(typedef, DEFAULT_TYPE_CONFLICT_HANDLER)
                );
            }

            case "IMAGE_CHPE_METADATA_X86": {
                StructureDataType S =
                    new StructureDataType("_IMAGE_CHPE_METADATA_X86", 0);
                S.add(ULONG, "Version", null);
                S.add(RVA,   "CHPECodeAddressRangeOffset", null);
                S.add(ULONG, "CHPECodeAddressRangeCount", null);
                S.add(RVA,   "WowA64ExceptionHandlerFunctionPointer", null);
                S.add(RVA,   "WowA64DispatchCallFunctionPointer", null);
                S.add(RVA,   "WowA64DispatchIndirectCallFunctionPointer", null);
                S.add(RVA,   "WowA64DispatchIndirectCallCfgFunctionPointer", null);
                S.add(RVA,   "WowA64DispatchRetFunctionPointer", null);
                S.add(RVA,   "WowA64DispatchRetLeafFunctionPointer", null);
                S.add(RVA,   "WowA64DispatchJumpFunctionPointer", null);
                S.setCategoryPath(getCategoryPath());

                TypedefDataType typedef =
                    new TypedefDataType("IMAGE_CHPE_METADATA_X86", S);
                return Optional.of(
                    typeManager.addDataType(typedef, DEFAULT_TYPE_CONFLICT_HANDLER)
                );
            }

            case "IMAGE_CHPE_METADATA_X86_V2": {
                DataType base = getType("IMAGE_CHPE_METADATA_X86").get();
                StructureDataType S =
                    new StructureDataType("_IMAGE_CHPE_METADATA_X86_V2", 0);
                S.add(base, "V1", null);

                S.add(RVA, "CompilerIATPointer", null);

                S.setCategoryPath(getCategoryPath());

                TypedefDataType typedef =
                    new TypedefDataType("IMAGE_CHPE_METADATA_X86_V2", S);
                return Optional.of(
                    typeManager.addDataType(typedef, DEFAULT_TYPE_CONFLICT_HANDLER)
                );
            }

            case "IMAGE_CHPE_METADATA_X86_V3": {
                DataType base = getType("IMAGE_CHPE_METADATA_X86_V2").get();
                StructureDataType S =
                    new StructureDataType("_IMAGE_CHPE_METADATA_X86_V3", 0);
                S.add(base, "V2", null);

                S.add(RVA, "WowA64RdtscFunctionPointer", null);

                S.setCategoryPath(getCategoryPath());

                TypedefDataType typedef =
                    new TypedefDataType("IMAGE_CHPE_METADATA_X86_V3", S);
                return Optional.of(
                    typeManager.addDataType(typedef, DEFAULT_TYPE_CONFLICT_HANDLER)
                );
            }

            case "IMAGE_ARM64EC_METADATA_CODE_RANGE_TYPE": {
                EnumDataType E =
                    new EnumDataType("IMAGE_ARM64EC_METADATA_CODE_RANGE_TYPE", U32.getLength());

                E.add("ARM64", 0, "Pure ARM64 code");
                E.add("ARM64EC", 1, "ARM64EC hybrid code (compatible with x64).");
                E.add("AMD64", 2, "x64 code");

                E.setCategoryPath(getCategoryPath());

                return Optional.of(
                    typeManager.addDataType(E, DEFAULT_TYPE_CONFLICT_HANDLER)
                );
            }

            case "IMAGE_ARM64EC_METADATA_CODE_RANGE": {
                StructureDataType S =
                    new StructureDataType("_IMAGE_ARM64EC_METADATA_CODE_RANGE", 0);

                S.setPackingEnabled(true);
                addBitField(S,
                    getType("IMAGE_ARM64EC_METADATA_CODE_RANGE_TYPE").get(),
                    2, "Type", null);

                addBitField(S, U32, 30, "RVA", null);
                S.add(U32, "Length", null);

                TypedefDataType typedef =
                    new TypedefDataType("IMAGE_ARM64EC_METADATA_CODE_RANGE", S);
                return Optional.of(
                    typeManager.addDataType(typedef, DEFAULT_TYPE_CONFLICT_HANDLER)
                );
            }

            case "IMAGE_ARM64EC_CODE_RANGE_ENTRY_POINT": {
                StructureDataType S =
                    new StructureDataType("_IMAGE_ARM64EC_CODE_RANGE_ENTRY_POINT", 0);

                S.add(RVA, "StartRva", null);
                S.add(RVA, "EndRva", null);
                S.add(RVA, "EntryPoint", null);

                TypedefDataType typedef =
                    new TypedefDataType("IMAGE_ARM64EC_CODE_RANGE_ENTRY_POINT", S);
                return Optional.of(
                    typeManager.addDataType(typedef, DEFAULT_TYPE_CONFLICT_HANDLER)
                );
            }

            case "IMAGE_ARM64EC_METADATA_REDIRECTION": {
                StructureDataType S =
                    new StructureDataType("_IMAGE_ARM64EC_METADATA_REDIRECTION", 0);

                S.add(RVA, "From", null);
                S.add(RVA, "To", null);

                TypedefDataType typedef =
                    new TypedefDataType("IMAGE_ARM64EC_METADATA_REDIRECTION", S);
                return Optional.of(
                    typeManager.addDataType(typedef, DEFAULT_TYPE_CONFLICT_HANDLER)
                );
            }

            case "IMAGE_ARM64_RUNTIME_FUNCTION_ENTRY": {
                StructureDataType S =
                    new StructureDataType("_IMAGE_ARM64_RUNTIME_FUNCTION_ENTRY", 0);

                S.add(RVA, "BeginAddress", null);

                StructureDataType T =
                    new StructureDataType("IMAGE_ARM64_RUNTIME_FUNCTION_ENTRY_PACKED", 0);
                {
                    T.setPackingEnabled(true);
                    addBitField(T, ULONG, 2, "Flag", null);
                    addBitField(T, ULONG, 11, "FunctionLength", null);
                    addBitField(T, ULONG, 3, "RegF", null);
                    addBitField(T, ULONG, 4, "RegI", null);
                    addBitField(T, ULONG, 1, "H", null);
                    addBitField(T, ULONG, 2, "CR", null);
                    addBitField(T, ULONG, 9, "FrameSize", null);
                    typeManager.addDataType(T, DEFAULT_TYPE_CONFLICT_HANDLER);
                }

                S.add(T, "packed", null);

                TypedefDataType typedef =
                    new TypedefDataType("IMAGE_ARM64_RUNTIME_FUNCTION_ENTRY", S);
                return Optional.of(
                    typeManager.addDataType(typedef, DEFAULT_TYPE_CONFLICT_HANDLER)
                );
            }

            case "IMAGE_ARM64_RUNTIME_FUNCTION_UNPACKED_ENTRY": {
                StructureDataType S =
                    new StructureDataType("IMAGE_ARM64_RUNTIME_FUNCTION_UNPACKED_ENTRY", 0);

                S.add(RVA, "BeginAddress", null);
                S.add(RVA, "ExceptionInfoRVA", null);

                return Optional.of(
                    typeManager.addDataType(S, DEFAULT_TYPE_CONFLICT_HANDLER)
                );
            }


            case "IMAGE_ARM64_RUNTIME_FUNCTION_ENTRY_XDATA": {
                StructureDataType S =
                    new StructureDataType("IMAGE_ARM64_RUNTIME_FUNCTION_ENTRY_XDATA", 0);

                S.setPackingEnabled(true);
                addBitField(S, ULONG, 18, "FunctionLength", null);
                addBitField(S, ULONG, 2, "Version", null);
                addBitField(S, ULONG, 1, "ExceptionDataPresent", null);
                addBitField(S, ULONG, 1, "EpilogInHeader", null);
                addBitField(S, ULONG, 5, "EpilogCount", null);
                addBitField(S, ULONG, 5, "CodeWords", null);

                return Optional.of(
                    typeManager.addDataType(S, DEFAULT_TYPE_CONFLICT_HANDLER)
                );
            }

            case "IMAGE_ARM64_RUNTIME_FUNCTION_ENTRY_XDATA_EXTENDED": {
                StructureDataType S =
                    new StructureDataType("IMAGE_ARM64_RUNTIME_FUNCTION_ENTRY_XDATA_EXTENDED", 0);

                S.setPackingEnabled(true);
                addBitField(S, ULONG, 16, "ExtendedEpilogCount", null);
                addBitField(S, ULONG, 8, "ExtendedCodeWords", null);

                return Optional.of(
                    typeManager.addDataType(S, DEFAULT_TYPE_CONFLICT_HANDLER)
                );
            }

            case "IMAGE_ARM64_RUNTIME_FUNCTION_EXTENDED_ENTRY": {
                StructureDataType S =
                    new StructureDataType("IMAGE_ARM64_RUNTIME_FUNCTION_EXTENDED_ENTRY", 0);
                S.add(getType("IMAGE_ARM64_RUNTIME_FUNCTION_ENTRY_XDATA").get(), "base", null);
                S.add(getType("IMAGE_ARM64_RUNTIME_FUNCTION_ENTRY_XDATA_EXTENDED").get(), "extension", null);

                return Optional.of(
                    typeManager.addDataType(S, DEFAULT_TYPE_CONFLICT_HANDLER)
                );
            }

            case "EpilogScope": {
                StructureDataType S =
                    new StructureDataType("EpilogScope", 0);

                S.setPackingEnabled(true);
                addBitField(S, ULONG, 18, "StartOffset", null);
                addBitField(S, ULONG, 4, "Res", null);
                addBitField(S, ULONG, 10, "ExceptionDataPresent", null);

                return Optional.of(
                    typeManager.addDataType(S, DEFAULT_TYPE_CONFLICT_HANDLER)
                );
            }
        }
        return super.getType(name);
    }

    @Override
    public CategoryPath getCategoryPath() {
        return super.getCategoryPath().extend("PE");
    }
}
